code/go/0chain.net/smartcontract/storagesc/clientpool.go

```diff
@@ -0,0 +1,722 @@
+package storagesc
+
+import (
+	"0chain.net/smartcontract"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/url"
+
+	cstate "0chain.net/chaincore/chain/state"
+	"0chain.net/chaincore/state"
+	"0chain.net/chaincore/transaction"
+	"0chain.net/core/common"
+	"0chain.net/core/datastore"
+	"0chain.net/core/util"
+)
+
+//
+// client read/write pool
+//
+
+func clientPoolKey(scKey, clientID string) datastore.Key {
+	return datastore.Key(scKey + ":clientpool:" + clientID)
+}
+
+type clientPool struct {
+	Allocations allocationPools `json:"allocations"`
+}
+
+// Encode pool
+// Implements util.Serializable interface.
+func (cPool *clientPool) Encode() []byte {
+	var b, err = json.Marshal(cPool)
+	if err != nil {
+		panic(err) // must never happens
+	}
+	return b
+}
+
+// Decode pool
+// Implements util.Serializable interface
+func (cPool *clientPool) Decode(p []byte) error {
+	return json.Unmarshal(p, cPool)
+}
+
+// Save pool in tree
+func (cPool *clientPool) save(ssContractKey, clientID string, balances cstate.StateContextI,
+) (err error) {
+	_, err = balances.InsertTrieNode(clientPoolKey(ssContractKey, clientID), cPool)
+	return
+}
+
+// todo: add description, rename it
+func (cPool *clientPool) blobberCut(aPoolID, bPoolID string, eTimestamp common.Timestamp,
+) []*allocationPool {
+	return cPool.Allocations.blobberCut(aPoolID, bPoolID, eTimestamp)
+}
+
+// Former removeEmpty
+// todo: add description
+// todo: clarify and rename - to remove allocation or allocations?
+func (cPool *clientPool) removeEmptyAllocation(aPoolID string, aPools []*allocationPool) {
+	cPool.Allocations.removeEmpty(aPoolID, aPools)
+}
+
+// For read operations only
+// todo: better rename it, adding "Read" in
+// todo: something probably wrong with it.
+// todo: clientPool entity should be used insted of some params to be passed?
+func (cPool *clientPool) moveBlobberCharge(ssContractKey string, sPool *stakePool,
```

Comment:
Might be best to use [function.go.](https://github.com/0chain/0chain/blob/master/code/go/0chain.net/smartcontract/storagesc/functions.go#L36) athough this is being redone in PR https://github.com/0chain/0chain/pull/317 where it has will be changed to [transferReward](https://github.com/0chain/0chain/blob/966d7479999ac15ce1642e029154e73c95405ba3/code/go/0chain.net/smartcontract/storagesc/functions.go#L16)